### PR TITLE
When xa is enabled, batch operations automatically support transactions

### LINKED ISSUE #20234: xa does not guarantee transactionality
## Bug Report
When the xa switch is turned on, two pieces of data are inserted into two datasources when inserting, but transactionality cannot be guaranteed.

### Which version of ShardingSphere did you use?
master

### Which project did you use? ShardingSphere-JDBC or ShardingSphere-Proxy?
ShardingSphere-Proxy

image.

### PR DESCRIPTION
**Pull request by natehuangting**
Fixes #20234 .

### CODE REVIEW

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java [hunk 1]
```diff
@@ -129,12 +135,33 @@ public ResponseHeader execute() throws SQLException {
...
+            if (null != transactionManager) {
+                transactionManager.commit();
+            }
+        } catch (final SQLException ex) {
+            if (null != transactionManager) {
+                transactionManager.rollback();
+            }
+            throw ex;
+        }
+        return responseHeader;
```
Comment (terrymanu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java:164]: Return value should name as `result`

**Review by terrymanu — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java [hunk 2]
```diff
@@ -129,12 +135,33 @@ public ResponseHeader execute() throws SQLException {
...
-        List result = proxySQLExecutor.execute(executionContext);
-        refreshMetaData(executionContext);
-        Object executeResultSample = result.iterator().next();
-        return executeResultSample instanceof QueryResult
-                ? processExecuteQuery(executionContext, result, (QueryResult) executeResultSample)
-                : processExecuteUpdate(executionContext, result);
+        TransactionStatus transactionStatus = backendConnection.getConnectionSession().getTransactionStatus();
+        SQLStatement sqlStatement = executionContext.getSqlStatementContext().getSqlStatement();
+        JDBCBackendTransactionManager transactionManager = null;
+        if (TransactionType.XA == transactionStatus.getTransactionType() && transactionStatus.isInTransaction()
```
Comment (strongduanmu) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java:141]: Can you exclude the scenario of routing to a single data node? At this time, using the local transaction can meet the requirements.

**Review by strongduanmu — CHANGES_REQUESTED** [review]

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java (hunk 2 above)
Comment (natehuangting) [reply shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java:141]: > Can you exclude the scenario of routing to a single data node? At this time, using the local transaction can meet the requirements.

@strongduanmu Sorry, I forgot to add the condition, I have changed it.

File: shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java [hunk 3]
```diff
@@ -127,12 +133,33 @@ public ResponseHeader execute() throws SQLException {
...
```
Comment (FlyingZC) [inline shardingsphere-proxy/shardingsphere-proxy-backend/src/main/java/org/apache/shardingsphere/proxy/backend/communication/jdbc/JDBCDatabaseCommunicationEngine.java:139]: Maybe not in the transaction `!transactionStatus.isInTransaction()` and then auto start the transaction.

**Comment by strongduanmu** [comment]
Hi @natehuangting, can you solve code conflict?

**Comment by strongduanmu** [comment]
This pr has been supported by #22029

**Pull request closed without merging**